Fusion workfile template builder

EmberLightVFX opened the pull request:
## Changelog Description
This PR starts the work to port Workfile Template Builder to Fusion!
I have gotten the creation of the placeholder nodes down with its data creation and so on.
Currently it can load footage but there are a lot more to fix and test until its ready for prime time.

My focus is to get everything up and running correctly and after clean up the code and the menu UI.
There are as for this PR creation a bunch of nuke code still in the fusion .py file. They will be converted to functional Fusion code while I'm working on this PR.

## Additional info
For the placeholder node I have created an "OPEmpty" fuse. It simply pushed the image along and have 0 controllers on it.
The placeholder-creator populates the fuse with the needed info on creation.
I took this path as it's waaaay easier and faster to search for a specific type of node in Fusion than taking an existing node, strip it of everything and then populate it with the needed info.
If there are any backlash I guess it would be done but this way is just easier for everyone.

## Testing notes:
1. Follow this tutorial to set everything up: https://www.youtube.com/watch?v=kfMJn3SXUGM
2. Try it out!

BigRoy commented on openpype/hosts/fusion/api/menu.py:
```diff
@@ -59,6 +67,17 @@ def __init__(self, *args, **kwargs):
+            "Build workfile from template...", self
+        )
+        create_placeholder_btn = QtWidgets.QPushButton(
+            "Creat placeholder...", self
```
```suggestion
            "Create placeholder...", self
```

BigRoy commented on openpype/hosts/fusion/api/menu.py:
```diff
@@ -59,6 +67,17 @@ def __init__(self, *args, **kwargs):
         manager_btn = QtWidgets.QPushButton("Manage...", self)
         libload_btn = QtWidgets.QPushButton("Library...", self)
+
+        build_workfile_from_template_btn = QtWidgets.QPushButton(
+            "Build workfile from template...", self
+        )
+        create_placeholder_btn = QtWidgets.QPushButton(
+            "Creat placeholder...", self
+        )
+        update_placeholder_btn = QtWidgets.QPushButton(
+            "Update placeholder...", self
+        )
```
Should we move this into a submenu or alike?

This is what it looks like in Maya:
![afbeelding](https://user-images.githubusercontent.com/2439881/228672352-2e518101-8a9a-4503-8728-291ba0c06f89.png)

Also like in Maya, let's move them to the bottom below reset framerange, etc.

EmberLightVFX replied:
Yep, that's the way to go but my first focus is to get everything up and running and after clean up the code and the menu UI. After that I'll change from draft to review :)

BigRoy reviewed (commented):
Some quick comments.

I noticed there are _tons_ of changes in here which are just cosmetic. E.g. renaming LoadSequnce to LoadClip, changing indentations on files that don't seemingly need to change for this PR. Admittedly these are changes I don't care too much about in this case but it's definitely making a code review more complicated.

I'd rather have a style change PR separate so that we can comment on that separately or something or just merge it quickly so it doesn't get too involved with bigger PRs like this.

BigRoy commented on openpype/hosts/fusion/api/workfile_template_builder.py:
```diff
@@ -0,0 +1,990 @@
+            offset (int): distance between copies
+        """
+
+        placeholder_node = nuke.toNode(placeholder.scene_identifier)
```
Hehe, I don't think this will work in Fusion. Is this still a draft PR or should I get to testing soon?

BigRoy commented on openpype/hosts/fusion/plugins/load/load_clip.py:
```diff
@@ -126,32 +130,31 @@ def loader_shift(loader, frame, relative=True):
 
-class FusionLoadSequence(load.LoaderPlugin):
-    """Load image sequence into Fusion"""
+class FusionLoadClip(load.LoaderPlugin):
```
Actually - I think this will be a major backwards compatible break. I believe the loader gets stored on the container and for that it stores the class's name as far as I know. I'm pretty sure this will break things.

It's imprinted [here](https://github.com/ynput/OpenPype/blob/5c8bbe28713aaabbabf99b4739f1654a0bea4b71/openpype/hosts/fusion/api/pipeline.py#L254) with the [value from e.g. here](https://github.com/ynput/OpenPype/blob/75ea09e238d684b6984e554fd95d0e524f4c036a/openpype/hosts/fusion/plugins/load/load_sequence.py#L167).

This will need to be reverted.

EmberLightVFX replied:
Thanks for pointing this out. Will revert back!

EmberLightVFX commented:
So finally the first complete revision is done!
You can create templates as expected that includes loaders and creators!
`workfile_template_builder.py` contains a bunch of TODO's still. Biggest TODO is to convert the `siblings` code (line 264-295 and 545-576 inc 2 functions each).
I don't fully understand what siblings does or when those parts would get executed so... Any help on those parts is highly appreciated!

EmberLightVFX commented:
Now it's time to start checking out the code if you want @BigRoy :)
There are a lot of duplicate codes between the classes FusionPlaceholderLoadPlugin and FusionPlaceholderCreatePlugin. Idk if it's possible to share it somehow? Or maybe it's just best to have them split if there are any need to modify per class later on.

EmberLightVFX commented:
@iLLiCiTiT I saw that you made most of the code for the Nuke implementation of `workfile_template_builder.py`
I would love to know what the siblings-code does in it to know if it's a needed part for the Fusion implementation and how I should code it up. :)

EmberLightVFX commented:
@friquette it seems like it was you who made the Nuke version. 
I would love to know what the siblings-code does in it to know if it's a needed part for the Fusion implementation and how I should code it up. :)

movalex commented:
Hey, thanks for your work, man!
Just in case, if you'd want to continue working on this, I updated this branch with the latest develop branch, resolved the conflicts and cleaned up all black formatting changes, so the diffs would be more clear. We can reformat the code later. 
Feel free to merge on top of this PR. 
https://github.com/ynput/OpenPype/compare/develop...movalex:OpenPype:feature/fusion-template-builder

EmberLightVFX commented:
> Hey, thanks for your work, man! Just in case, if you'd want to continue working on this, I updated this branch with the latest develop branch, resolved the conflicts and cleaned up all black formatting changes, so the diffs would be more clear. We can reformat the code later. Feel free to merge on top of this PR. [develop...movalex:OpenPype:feature/fusion-template-builder](https://github.com/ynput/OpenPype/compare/develop...movalex:OpenPype:feature/fusion-template-builder)

Thanks @movalex !
I currently don't have much time to work on this integration so if you wish to take over the development of it I'm more than happy to do so!
